### ISSUE TITLE
Investigate loss function with fewer random sampling points

### ISSUE THREAD

**Issue by JordiBolibar** [ODINN-SciML/ODINN.jl#22]
There's the hypothesis that computing the loss with fewer points, instead of using the full `H` or `V` matrix can result in an improved performance during backpropagation. We should test this and see if this also can act as regularization. It would be nice to come up with a sweet spot between performance and accuracy.

**Comment by JordiBolibar** [comment]
I have made a benchmark to test the different loss functions based on `H`, `V` and `HV` with the full matrix or a subset of random values within the glacier. Here are the results:

**`H`:**  
_Full matrix:_ 203.710 s (3655336127 allocations: 188.45 GiB) 
		      209.479 s (3650073161 allocations: 188.16 GiB)
_50 samples:_ 202.561 s (3737450206 allocations: 191.94 GiB)
_10 samples:_ 220.600 s (3837529412 allocations: 197.03 GiB)

**`V`:**  
_Full matrix:_ 519.855 s (8589977971 allocations: 471.25 GiB) 
		      503.255 s (7985655295 allocations: 437.93 GiB)
_10 samples:_ 301.433 s (5164587956 allocations: 275.87 GiB)

**`HV`:**  
_Full matrix:_ 480.228 s (7931932251 allocations: 434.23 GiB) 
_10 samples:_ 396.731 s (6893550375 allocations: 372.17 GiB)

The results are interesting, and I still haven't been able to interpret them. It seems like the fastest solution is the loss function using `H`, but for that case there is no speed-up when using only a subset of the matrix.

Surprisingly, using both `H` and `V` for the loss (`HV`) is faster than just using `V`. And for the cases where `V` is involved, all the runtimes are much slower than with only `H`. Luckily, for those cases there seems to be a speed-up when using only a subset of the matrix. I still haven't been able to understand this behaviour, I have no clue why this should only work for a certain loss function. 

This should be further investigated, together with #33.

**Comment by JordiBolibar** [comment]
OK, I think I understand the differences for the two loss functions. For `H`, for which there’s no speed-up, it’s possibly because all computations in the forward pass are already performed on the full matrix, so if the loss is only computed for a subset of points, this leads to no improvement, because all information is already available. However, for `V`, where there is an improvement, I need to perform an extra operation in order to get `V` from `H`. If I choose a subset of points of the matrix, then this extra operation to get `V` is cheaper, leading to a faster gradients.

In a nutshell: this is only worth it if it involves extra computations on top of what is done in the forward pass.

**Comment by JordiBolibar** [comment]
@facusapienza21 I think it might a good idea to get rid of this? This would greatly reduce the amount of code in the loss function, and seeing that it doesn't speed things up, it is just simpler to compute the loss on the whole matrix.

**Comment by facusapienza21** [comment]
I agree. It is good we have this issue closed for future reference but this doesn't seem to be working. Maybe we can put a tag or a commit message indicating the last version for which we still have the sampling implemented. This could be useful if at some point in the future we want to try this again, maybe with some difference respect to the loss function, etc.

**Comment by JordiBolibar** [comment]
OK, I'll create a separate branch after the big PR that I'm preparing.

**Comment by JordiBolibar** [comment]
Closing this stale issue, since it doesn't seem to be relevant anymore. We can always open a new one later on.
